File: 01.patterns/pattern-08.js
// this will output this
// * * * * * * * * * 
//   * * * * * * *
//     * * * * *
//       * * *
//         *

// we can actually see the pattern here.. 
// [0, 9, 0]
// [1, 7, 1]
// [2, 5, 2]
// [3, 3, 3]
// [4, 1, 4]
// for conection with rows from column
// we can see that there is actually reverse of it, which is basically n for spaces
// we can see that for the stars it is actually 2 * n - (2i+1)

let n = 5
for(let i = 0; i < n; i++){
    let stars = ""
    let space1 = ""
    let space2 = ""
    for(let j = 0; j < 2*n - (2*i + 1); j++){
        stars += "* "
    }

    for(let j = 0; j < i; j++ ){
        space1 += "  "
        space2 += "  "
    }

    console.log(space1 + stars + space2)
}

console.log("--- revision starts here ---")

// * * * * * * * * * 
//   * * * * * * *
//     * * * * *
//       * * *
//         *

let k = 5
for(let i = 0; i < k; i++){
    let stars = ""
    for(let j = 0; j < 2 * k - 2 * i - 1; j++){
        stars += "* "
    }

    let space = ""
    for(let n = 0; n < i; n++){
        space += "  "
    }
    console.log(space + stars)
}

// final revision
// * * * * * * * * * 
//   * * * * * * *
//     * * * * *
//       * * *
//         *


let rows = 5

for(let i = 0; i < rows; i++){
    let space = ''
    for(let j = 0; j < i; j++){
        space += '  '
    }
    let stars = ''
    for(let j = 0; j < 2 * rows - (2 * i + 1); j++){
        stars += '* '
    }

    console.log(space + stars)
}

// practise for a job application
console.log("-----for job application-----")

// * * * * * * * * * 
//   * * * * * * *
//     * * * * *
//       * * *
//         *

let p = 5

for(let i = 0; i < p; i++){
    let space = ''
    for(let j = 0; j < i; j++){
        space += '  '
    }

    let stars = ''
    
    for(let j = 0; j < 2 * p - (2 * i + 1); j++){
        stars += '* '
    }

    console.log(space + stars + space)
}

File: 01.patterns/pattern-08.go
// ------------------ revision with js and go -------------------------
package main

import (
	"fmt"
	"strings"
)

// * * * * * * * * *
//   * * * * * * *
//     * * * * *
//       * * *
//         *

// 0 9 0
// 1 7 1
// 2 5 2
// 3 3 3
// 4 1 4

func main() {
	p := 5

	// // for i := 0; i < p; i++ {

	// // }

	// for i := 0; i < p; i++ {
	// 	space := strings.Repeat("  ", p-i-1)
	// 	stars := strings.Repeat("* ", 2*i+1)
	// 	fmt.Println(space + stars)
	// }

	for i := 0; i < p; i++ {
		space := strings.Repeat("  ", i)
		stars := strings.Repeat("* ", 2*p-(2*i+1))
		fmt.Println(space + stars)
	}

	fmt.Println("---------")

	for i := 0; i < p; i++ {
		var b strings.Builder
		b.Grow(2*(2*p) - 1)
		for j := 0; j < i; j++ {
			b.WriteString("  ")
		}

		for j := 0; j < 2*p-(2*i+1); j++ {
			b.WriteString("* ")
		}

		fmt.Println(b.String())
	}
}
